File: java/commonWeb/src/main/java/web/template/controller/common/IndexController.java
package web.template.controller.common;
import java.awt.Image;
import java.awt.Rectangle;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import javax.imageio.ImageIO;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.apache.commons.io.IOUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.util.FileCopyUtils;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;
import com.txj.common.FileHelper;
import com.txj.common.ImageHandleHelper;
import com.txj.common.ThreadHelper;
import com.txj.common.entity.Result;

import web.template.intf.IRealTimeInitService;
import web.template.intf.IUserService;
@RestController
@RequestMapping("/index")
public class IndexController extends BaseController {

	@Autowired
	private IUserService userService;
	
	@Autowired
	private IRealTimeInitService realTimeInitService;
	
	/**
	 * 等待池表
	 */
	public Set<String> waitPoolSet;
	
	/**
	 * 用户和等待池表，确保每个用户只能在同一个等待池上等待。
	 */
	public final static HashSet<String> USERNAME_AND_POOL_SET;

	static {
		USERNAME_AND_POOL_SET = new HashSet<String>();
	}

	public IndexController() {
		waitPoolSet = new HashSet<String>();
		waitPoolSet.add("newsAlarm");
	}

	/**
	 * 实时挂载线程到等待池
	 * @param realTimePool	
	 * @param realTimeVersion
	 * @param username
	 * @return
	 */
	private final Result realTime(final String realTimePool, final String realTimeVersion, final String username) {
		if (waitPoolSet.contains(realTimePool)) {
			final String[] newestVersion = new String[1];
			boolean initRet = ThreadHelper.compareControllerVersion(realTimePool, realTimeVersion, newestVersion);
			if (realTimeVersion == null) {
				initRet = realTimeInitService.init(realTimePool, username);
			}
			final Map<String, String> data = new HashMap<String, String>();
			if (initRet) {
				final String usernameAndPoolKey = username + realTimePool;
				if (USERNAME_AND_POOL_SET.contains(usernameAndPoolKey)) {
					return new Result(-1, "当前用户已有线程在等待池中，无法向等待池放入新线程。", null);
				} else {
					synchronized (USERNAME_AND_POOL_SET) {
						if (USERNAME_AND_POOL_SET.contains(usernameAndPoolKey)) {
							return new Result(-1, "当前用户已有线程在等待池中，无法向等待池放入新线程。", null);
						} else {
							USERNAME_AND_POOL_SET.add(usernameAndPoolKey);
						}
					}
				}
				ThreadHelper.batchWait(realTimePool, 60000);
				synchronized (USERNAME_AND_POOL_SET) {
					USERNAME_AND_POOL_SET.remove(usernameAndPoolKey);
				}
				initRet = ThreadHelper.compareControllerVersion(realTimePool, realTimeVersion, newestVersion);
				data.put("realTimePool", realTimePool);
				data.put("realTimeVersion", newestVersion[0]);
				return new Result(1, null, data);
			} else {
				data.put("realTimePool", realTimePool);
				data.put("realTimeVersion", newestVersion[0]);
				return new Result(0, null, data);
			}
		} else {
			return new Result(-1, "该等待池未开放。", null);
		}
	}

	/**
	 * 
	 * @param realTimePool
	 * @param realTimeVersion
	 * @param signKey
	 * @return
	 */
	@RequestMapping("/anonymousRealTime")
	public final Result anonymousRealTime(final String realTimePool, final String realTimeVersion, final String signKey) {
		return realTime(realTimePool, realTimeVersion,
				/* systemService.LoadUsernameBySignKey(signKey) */"zhang");
	}

	/**
	 * 
	 * @param realTimePool
	 * @param realTimeVersion
	 * @return
	 */
	@RequestMapping("/realTime")
	public final Result realTime(String realTimePool, String realTimeVersion) {
		return realTime(realTimePool, realTimeVersion, username());
	}

	/**
	 * 加载地区选择json
	 * 
	 * @return
	 */
	public Result areaSelect() {
		return null;
	}

	/**
	 * 登陆后加载基础数据
	 * 
	 * @return
	 */
	@RequestMapping("/loadLoginData")
	public Result loadLoginData() {
		Map<String, Object> data = new HashMap<String, Object>();
		data.put("leftMenus", userService.loadLeftMenus(username()));
		return new Result(0, null, data);
	}

	/**
	 * 获取左侧菜单
	 * 
	 * @return
	 */
	@RequestMapping("/loadLeftMenus")
	public Result loadLeftMenus() {
		return new Result(0, "", userService.loadLeftMenus(username()));
	}

	/**
	 * 显示图片
	 * 
	 * @param imgName
	 */
	@RequestMapping("/showImage")
	public Result showImage(String imgName, HttpServletRequest request, HttpServletResponse response) {
		InputStream inputStream;
		try {
			File imgFile = realFile("/WEB-INF/uploadFiles/" + imgName);
			if (imgFile.exists()) {
				inputStream = new FileInputStream(imgFile);
				response.addHeader("Cache-control", "max-age=" + Integer.MAX_VALUE);
				response.setContentType("image/jpeg");
				FileCopyUtils.copy(inputStream, response.getOutputStream());
			} else {
				response.setStatus(404);
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
		return null;
	}

	/**
	 * 下载文件
	 * 
	 * @param pathName
	 * @param fileName
	 * @param fileDesc
	 */
	@RequestMapping("/downFile")
	public void downFile(String fileName, String fileDesc, HttpServletResponse response) {
		File file = realFile("/WEB-INF/uploadFiles/" + fileName);
		try (FileInputStream fileInputStream = new FileInputStream(file)) {
			response.setContentType("application/octet-stream");
			response.setHeader("Content-disposition",
					"attachment; filename=" + new String(fileDesc.getBytes("utf-8"), "ISO8859-1"));
			IOUtils.copy(fileInputStream, response.getOutputStream());
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	/**
	 * 切割单张图片
	 * 
	 * @param imgName
	 *            图片名称
	 * @param imgWidth
	 *            图片宽度
	 * @param imgHeight
	 *            图片高度
	 * @param x
	 *            左上角切点x坐标
	 * @param y
	 *            左上角切点y坐标
	 * @param w
	 *            切割宽度
	 * @param h
	 *            切割高度
	 * @return 返回切割结果，thumbnailName（切割图）、imgName（缩略图）、
	 */
	@RequestMapping("/singleImageCrop")
	public Result singleImageCrop(String imgName, int imgWidth, int imgHeight, int x, int y, int w,
			int h, HttpServletRequest request) {
		try {
			File file = realFile("/WEB-INF/uploadFiles/" + imgName);
			if (file.exists()) {
				Image scaleImage = ImageHandleHelper.scale(ImageIO.read(file), imgWidth, imgHeight);
				Image cutImage = ImageHandleHelper.cutPicByRect(scaleImage, new Rectangle(x, y, w, h));
				Image thumbnailImg = ImageHandleHelper.scale(cutImage, 150,
						cutImage.getHeight(null) * 150 / cutImage.getWidth(null));
				Map<String, String> data = new HashMap<String, String>();
				data.put("imgName", FileHelper.SaveImageBySha1(cutImage, realPath("/WEB-INF/uploadFiles/")));
				data.put("thumbnailName",
						FileHelper.SaveImageBySha1(thumbnailImg, realPath("/WEB-INF/uploadFiles/")));
				return new Result(0, null, data);
			} else {
				return new Result(-1, "图片不存在，切割失败！", null);
			}
		} catch (IOException e) {
			e.printStackTrace();
			return new Result(-1, e.getMessage(), null);
		}
	}

	/**
	 * 上次单张图片
	 * 
	 * @param fileUpload
	 *            上传图片
	 * @param request
	 *            请求对象
	 * @param pathName
	 *            路径名称
	 * @return
	 */
	@RequestMapping("/uploadSingleImage")
	public Result uploadSingleImage(@RequestParam("fileUpload") MultipartFile fileUpload, HttpServletRequest request) {
		try {
			File target = realFile("/WEB-INF/uploadFiles/");
			String sha1 = FileHelper.SaveInputStreamBySha1(fileUpload.getInputStream(), target.getAbsolutePath());
			Image image = ImageIO.read(new File(target.getAbsoluteFile() + File.separator + sha1));
			Image scaleImg = ImageHandleHelper.scale(image, 150,
					image.getHeight(null) * 150 / image.getWidth(null));
			String thumbnailName = FileHelper.SaveImageBySha1(scaleImg, target.getAbsolutePath());
			Map<String, Object> dataMap = new HashMap<>();
			dataMap.put("thumbnailName", thumbnailName);
			dataMap.put("imgName", sha1);
			dataMap.put("imgWidth", image.getWidth(null));
			dataMap.put("imgHeight", image.getHeight(null));
			return new Result(0, null, dataMap);
		} catch (IllegalStateException e) {
			e.printStackTrace();
			throw new RuntimeException(e);
		} catch (IOException e) {
			e.printStackTrace();
			throw new RuntimeException(e);
		}
	}

	/**
	 * 上传多个文件
	 * 
	 * @param fileUploads
	 *            上传的文件
	 * @param pathName
	 *            路径名称
	 * @return
	 */
	@RequestMapping("/uploadFiles")
	public Result uploadFiles(@RequestParam("fileUploads") MultipartFile fileUploads, HttpServletRequest request) {
		try {
			File target = realFile("/WEB-INF/uploadFiles/");
			return new Result(0, null,
					FileHelper.SaveInputStreamBySha1(fileUploads.getInputStream(), target.getAbsolutePath()));
		} catch (IOException e) {
			e.printStackTrace();
			return new Result(-1, e.getMessage(), null);
		}
	}
}

File: c#/CommonWeb/WebApplication1/Controllers/Common/IndexController.cs
using CommonHelper.Helper;
using CommonHelper.Helper.CommonEntity;
using CommonWeb.Filter.Common;
using CommonWeb.Intf;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Web;
using System.Web.Mvc;
namespace WebApplication1.Controllers.Common
{
    /// <summary>
    /// 通用控制器
    /// </summary>
    public sealed partial class IndexController : FastController
    {
        /// <summary>
        /// 实时刷新的业务处理类
        /// </summary>
        public IRealTimeInitService RealTimeInitService { set; get; }

        /// <summary>
        /// 用户专用的接口类，提供用户操作的业务逻辑
        /// </summary>
        public IUserService UserService { set; get; }

        /// <summary>
        /// 系统业务接口类，提供系统的信息
        /// </summary>
        public ISystemService SystemService { set; get; }

        /// <summary>
        /// 用户和等待池表，确保每个用户只能在同一个等待池上等待。
        /// </summary>
        public static HashSet<string> UsernameAndPoolSet { set; get; }

        static IndexController()
        {
            UsernameAndPoolSet = new HashSet<string>();
        }

        /// <summary>
        /// 加载登陆后数据
        /// </summary>
        /// <returns></returns>
        [Compress]
        public JsonResult LoadLoginData()
        {
            Dictionary<string, object> data = new Dictionary<string, object>();
            data.Add("leftMenus", UserService.LoadLeftMenus(User.Identity.Name));
            data.Add("username", User.Identity.Name);
            return MyJson(new Result { code = 0, data = data });
        }

        /// <summary>
        /// 匿名的实时检测最新版本号，如果发现版本号发生变化，则马上
        /// 返回，如果没变化，则等待30秒或30秒内版本号发生变化后返回
        /// （需要传入认证参数：createDate（yyyyMMddHHmmss）、r（随机字符串）、）
        /// </summary>
        /// <returns></returns>
        [AllowAnonymous]
        [Sign(new string[] { "realTimePool", "realTimeVersion", "createDate", "r" })]
        public JsonResult AnonymousRealTime(string realTimePool, string realTimeVersion,string signKey)
        {
            return RealTime(realTimePool, realTimeVersion,UserService.LoadUsernameBySignKey(signKey));
        }

        /// <summary>
        /// 实时刷新
        /// </summary>
        /// <param name="realTimePool"></param>
        /// <param name="realTimeVersion"></param>
        /// <param name="username"></param>
        /// <returns></returns>
		private JsonResult RealTime(string realTimePool,string realTimeVersion,string username)
		{
			if (RealTimeInitService.GetWaitPoolSet().Contains(realTimePool))
            {
                string newestVersion;
                bool initRet = ThreadHelper.CompareControllerVersion(realTimePool, realTimeVersion, out newestVersion);
                if (string.IsNullOrEmpty(realTimeVersion))
                {
                    initRet = RealTimeInitService.Init(realTimePool, username);
                }
                if (initRet)
                {
                    string usernameAndPoolKey = username + realTimePool;
                    if (UsernameAndPoolSet.Contains(usernameAndPoolKey))
                    {
                        return MyJson(new Result { code = -1, msg = "当前用户已有线程在等待池中，无法向等待池放入新线程。" });
                    }
                    else
                    {
                        lock (UsernameAndPoolSet)
                        {
                            if (UsernameAndPoolSet.Contains(usernameAndPoolKey))
                            {
                                return MyJson(new Result { code = -1, msg = "当前用户已有线程在等待池中，无法向等待池放入新线程。" });
                            }
                            else
                            {
                                UsernameAndPoolSet.Add(usernameAndPoolKey);
                            }
                        }
                    }
                    ThreadHelper.BatchWait(realTimePool, 60000);
                    lock (UsernameAndPoolSet)
                    {
                        UsernameAndPoolSet.Remove(usernameAndPoolKey);
                    }
					initRet = ThreadHelper.CompareControllerVersion(realTimePool, realTimeVersion, out newestVersion);
                    return MyJson(new Result { code = (short)(initRet?1:0),data=new Dictionary<string, string>
                        {
                            ["realTimePool"]= realTimePool,
                            ["realTimeVersion"]= newestVersion
                        }
                    });
                }
                else
                {
                    return MyJson(new Result { code = 0 ,data = new Dictionary<string, string>
                        {
                            ["realTimePool"] = realTimePool,
                            ["realTimeVersion"] = newestVersion
                        }
                    });
                }
            }
            else
            {
                return MyJson(new Result { code = -1,msg = "该等待池未开放。" });
            }
		}

        /// <summary>
        /// 实时检测最新版本号，如果发现版本号发生变化，则马上
        /// 返回，如果没变化，则等待30秒或30秒内版本号发生变化后返回
        /// </summary>
        /// <param name="realTimePool">等待池名称</param>
        /// <param name="realTimeVersion">版本号</param>
        /// <returns>返回结果</returns>
        public JsonResult RealTime(string realTimePool,string realTimeVersion)
        {
			return RealTime(realTimePool,realTimeVersion,User.Identity.Name);   
        }

        /// <summary>
        /// 上传单张图片
        /// </summary>
        /// <returns></returns>
        [SizeFilter(3145728, "上传文件大小不能超过3M，可通过压缩减少文件大小。")]
        [OperCount(CountLimit = 10,ClearMillisecond = 60000)]
        public JsonResult UploadSingleImage(HttpPostedFileBase fileUpload)
        {
            Image image = null;
            try
            {
                image = Image.FromStream(fileUpload.InputStream);
            }
            catch
            {
                return MyJson(new Result { code = -1, msg = "上传的文件可能不是图片文件，无法处理。" });
            }
            try
            {
                string imgName = FileHelper.SaveImageBySha1(image, SystemService.UploadFilePath());
                using (Image img = Image.FromFile($"{SystemService.UploadFilePath()}{s}{imgName}"))
                {
                    using (Image thumbnailImg = img.GetThumbnailImage(150, img.Height * 150 / img.Width, () => false, IntPtr.Zero))
                    {
                        string thumbnailName = FileHelper.SaveImageBySha1(thumbnailImg, SystemService.UploadFilePath());
                        return MyJson(new Result { code = 0, data = new { thumbnailName = thumbnailName, imgName = imgName, imgWidth = img.Width, imgHeight = img.Height } });
                    }
                }
            }
            catch (Exception ex)
            {
                return MyJson(new Result { code = -1, msg = ex.Message });
            }
        }

        /// <summary>
        /// 多个文件上传
        /// </summary>
        /// <param name="fileUploads"></param>
        /// <param name="pathName"></param>
        /// <returns></returns>
        [SizeFilter(52428800, "上传文件大小不能超过50M，可通过压缩减少文件大小。")]
        [OperCount(CountLimit = 100, ClearMillisecond = 60000)]
        public JsonResult UploadFiles(HttpPostedFileBase fileUploads)
        {
            return MyJson(new Result { code = 0, data = FileHelper.SaveFileNameBySha1(fileUploads.InputStream, SystemService.UploadFilePath()) });
        }

        /// <summary>
        /// 切割单张图片
        /// </summary>
        /// <param name="imgName">图片名称</param>
        /// <param name="imgWidth">图片宽度</param>
        /// <param name="imgHeight">图片高度</param>
        /// <param name="x">左上角切点x坐标</param>
        /// <param name="y">左上角切点y坐标</param>
        /// <param name="w">切割宽度</param>
        /// <param name="h">切割高度</param>
        public JsonResult SingleImageCrop(string imgName, int imgWidth, int imgHeight, int x, int y, int w, int h)
        {
            string imgPath = $"{SystemService.UploadFilePath()}{s}{imgName}";
            if (System.IO.File.Exists(imgPath))
            {
                Bitmap bitmap = null;
                try
                {
                    bitmap = new Bitmap(imgPath);
                }
                catch
                {
                    return MyJson(new Result { code = -1, msg = "该文件可能不是图片文件，切割失败！" });
                }
                try
                {
                    using (bitmap)
                    {
                        using (Image cutImg = ImageHandleHelper.CutPicByRect(bitmap.GetThumbnailImage(imgWidth, imgHeight, () => false, IntPtr.Zero), x, y, w, h))
                        {

                            using (Image thumbnailImg = cutImg.GetThumbnailImage(150, cutImg.Height * 150 / cutImg.Width, () => false, IntPtr.Zero))
                            {
                                return MyJson(new Result
                                {
                                    code = 0,
                                    data = new
                                    {
                                        thumbnailName = FileHelper.SaveImageBySha1(thumbnailImg, SystemService.UploadFilePath()),
                                        imgName = FileHelper.SaveImageBySha1(cutImg, SystemService.UploadFilePath())
                                    }
                                });
                            }
                        }
                    }
                }
                catch (Exception ex)
                {
                    return MyJson(new Result { code = -1, msg = ex.Message });
                }
            }
            else
            {
                return MyJson(new Result { code = -1, msg = "图片不存在，切割失败！" });
            }
        }

		/// <summary>
        /// 允许匿名访问的显示上传的图片
        /// </summary>
        /// <param name="pathName">图片路径名称</param>
        /// <param name="imgName">图片名称</param>
        [AllowAnonymous]
        [Sign(new string[] { "createDate", "r" ,"imgName" })]
        [OutputCache(Duration = int.MaxValue)]
        public void AnonymousShowImage(string imgName)
        {
            ShowImage(imgName);
        }

        /// <summary>
        /// 显示上传的图片
        /// <param name="pathName">图片路径名称</param>
        /// <param name="imgName">图片名称</param>
        /// </summary>
        [OutputCache(Duration = int.MaxValue)]
        public void ShowImage(string imgName)
        {
            Response.ContentType = "image/jpeg";
            string imgPath = $"{SystemService.UploadFilePath()}{s}{imgName}";
            if (System.IO.File.Exists(imgPath))
            {
                using (Stream stream = System.IO.File.OpenRead(imgPath))
                {
                    StreamHelper.CopyStream(stream, Response.OutputStream, false);
                }
            }
            else
            {
                Response.StatusCode = 404;
            }
        }

        /// <summary>
        /// 下载文件
        /// </summary>
        /// <param name="pathName">文件路径名称</param>
        /// <param name="fileName">文件sha1名称</param>
        /// <param name="fileDesc">文件的中文描述</param>
        /// <returns></returns>
        [OutputCache(Duration = int.MaxValue)]
        public FilePathResult DownFile(string fileName, string fileDesc)
        {
            string filePath = $"{SystemService.UploadFilePath()}{s}{fileName}";
            if (System.IO.File.Exists(filePath))
            {
                return File(filePath, MimeMapping.GetMimeMapping(filePath), fileDesc);
            }
            else
            {
                Response.StatusCode = 404;
                return null;
            }
        }
    }
}